### DISCUSSION TITLE
Transparent window glfw window lagging

### DISCUSSION THREAD

**Discussion by ItsVoltz** [AllenDang/giu#137] [Q&A] [answered]
So I was taking a look at nvtool's source and using Nicify's WindowMove hook. However it is really laggy when updating the GLFW Window. Any ideas how to fix this? I'm attempting to just achieve frameless windows that can be dragged flawlessly, I haven't worked with glfw before so I have no clue what I'm doing. Any help and explanations appreciated

https://user-images.githubusercontent.com/29076540/106997639-38546900-6751-11eb-9b42-d1999e7ad4bf.mp4

My example window code:

```go
func main() {
	wnd := g.NewMasterWindow("Transparent Test", 180, 80, g.MasterWindowFlagsNotResizable|g.MasterWindowFlagsFrameless|g.MasterWindowFlagsTransparent, nil)
	wnd.SetBgColor(color.RGBA{R: 0, G: 0, B: 0, A: 0,})

	platform := g.Context.GetPlatform().(*imgui.GLFW) // from nvtool
	glfwWindow = platform.GetWindow() // from nvtool
	applyWindowProperties(glfwWindow) // from nvtool

	wnd.Run(func() {
		UseWindowMove(glfwWindow, mwDragArea, &mwMoveState) // nvtool move hook
		g.Window("Transparent Test").Layout(
			g.Line(
				g.Button("Button").OnClick(func() {
					fmt.Println("button clicked")
				}),
				g.Button("Close").OnClick(func() {
					os.Exit(0)
				}),
			))
	})
}
```

Nicify's code:

```go
type MWMoveState struct {
	moveable   bool
	prevMouseX int
	prevMouseY int
}

var mwMoveState = MWMoveState{}
var mwDragArea  = image.Rectangle{image.ZP, image.Point{180, 30}}

func applyWindowProperties(window *glfw.Window) {
	hwnd := win.HWND(unsafe.Pointer(glfwWindow.GetWin32Window()))
	win.SetWindowCompositionAttribute(hwnd, 3, 0, 0, 0)
	g.Context.GetPlatform().SetTPS(144)
	glfwWindow.SetFocusCallback(func(w *glfw.Window, focused bool) {
		if focused {
			glfwWindow.SetOpacity(0.98)
			return
		}
		glfwWindow.SetOpacity(1)
	})
}

func UseWindowMove(glfwWindow *glfw.Window, dragArea image.Rectangle, mwMoveState *MWMoveState) {
	mouseX, mouseY := glfwWindow.GetCursorPos()
	prevPosX, prevPosY := glfwWindow.GetPos()

	if g.IsMouseClicked(0) {
		clickPoint := image.Point{X: int(mouseX / float64(imgui.DPIScale)), Y: int(mouseY / float64(imgui.DPIScale))}
		mwMoveState.moveable = clickPoint.In(dragArea)
		mwMoveState.prevMouseX = int(mouseX)
		mwMoveState.prevMouseY = int(mouseY)
	}
	if mwMoveState.moveable && g.IsMouseDown(0) {
		offsetX := int(mouseX) - mwMoveState.prevMouseX
		offsetY := int(mouseY) - mwMoveState.prevMouseY
		glfwWindow.SetPos(prevPosX+int(offsetX), prevPosY+int(offsetY))
	}
}
```

EDIT:

it seems to work fine with single window?
Any idea how to get it to work properly Multi windows (g.Window())

https://user-images.githubusercontent.com/29076540/106999052-836f7b80-6753-11eb-8084-9a887ee23f42.mp4

**Comment by Rel1cx** [comment, marked as answer]
The problem with glfw window lagging is that giu's default event pulling rate is only 60 ticks per second, which is not enough for high frequency operations like dragging. You can simply set it to 120 ticks per second or higher by calling `g.Context.GetPlatform().SetTPS(120)`

**Reply by ItsVoltz** [reply]
Thank you, maybe you know the answer to the other issue since I am using some of your functions you use in your project. https://github.com/AllenDang/giu/discussions/139. Any idea how to fix the weird transparency on like text and the corners of widgets?
